### PR TITLE
[EWS] Add CheckOutPullRequest step

### PR DESCRIPTION
**Pull request by JonWBedard**
#### b28a364ba6da2297d556b9aecd2eb32d181b5331
<pre>
[EWS] Add CheckOutPullRequest step
<a href="https://bugs.webkit.org/show_bug.cgi?id=235157">https://bugs.webkit.org/show_bug.cgi?id=235157</a>
&lt;rdar://problem/87487209  &gt;

Reviewed by Aakash Jain.

* Tools/CISupport/ews-build/factories.py:
(Factory.__init__): Add CheckOutPullRequest step.
(StyleFactory.__init__): Ditto.
(WatchListFactory.__init__): Ditto.
* Tools/CISupport/ews-build/factories_unittest.py:
(TestExpectedBuildSteps): Add CheckOutPullRequest step.
* Tools/CISupport/ews-build/steps.py:
(ApplyPatch.doStepIf): Only run step if a patch is provided.
(ApplyPatch.hideStepIf): Only display step if a patch was provided.
(ApplyPatch.getResultSummary): Add SKIPPED description.
(CheckOutPullRequest):
(CheckOutPullRequest.__init__):
(CheckOutPullRequest.doStepIf): Only run step if a pull request is provided.
(CheckOutPullRequest.hideStepIf): Only display step if a pul request was provided.
(CheckOutPullRequest.run): Configure new remote, fetch remote and checkout pull request
branch from the newly configured remote.
(CheckOutPullRequest.getResultSummary):
* Tools/CISupport/ews-build/steps_unittest.py:

Canonical link: <a href="https://commits.webkit.org/246230@main">https://commits.webkit.org/246230@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@288314">https://svn.webkit.org/repository/webkit/trunk@288314</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

### CODE REVIEW

**Comment by JonWBedard** [comment]
Some examples of this working:
    Applying a pull-request: https://ews-build.webkit-uat.org/#/builders/1/builds/1605
    Applying a patch: https://ews-build.webkit-uat.org/#/builders/1/builds/1603

File: Tools/CISupport/ews-build/steps.py [hunk 1]
```diff
@@ -405,49 +405,88 @@ def evaluateCommand(self, cmd):
...
-        d.addCallback(lambda res: shell.ShellCommand.start(self))
+        d.addCallback(lambda res: steps.ShellSequence.run(self))
+        return d
+
+    def run_github(self):
+        remote = self.getProperty('github.head.repo.full_name', 'origin').split('/')[0]
+        project = self.getProperty('github.head.repo.full_name', self.getProperty('project'))
+        branch = self.getProperty('github.head.ref', 'main')
+
+        self.commands.append(util.ShellArg(
```
Comment (aj062) [inline Tools/CISupport/ews-build/steps.py]: Let's have all these commands defined in a list at one place, and then loop over them.

File: Tools/CISupport/ews-build/steps.py [hunk 2]
```diff
@@ -405,49 +405,76 @@ def evaluateCommand(self, cmd):
...
+        d.addCallback(lambda res: steps.ShellSequence.run(self))
+        return d
+
+    def run_github(self):
+        remote = self.getProperty('github.head.repo.full_name', 'origin').split('/')[0]
+        project = self.getProperty('github.head.repo.full_name', self.getProperty('project'))
+        branch = self.getProperty('github.head.ref', 'main')
+
+        for command in [
+            ['/bin/sh', '-c', 'git remote add {} {}{}.git & true'.format(remote, GITHUB_URL, project)],
```
Comment (aj062) [inline Tools/CISupport/ews-build/steps.py]: `& true` seems to be for debugging

Comment (JonWBedard) [reply Tools/CISupport/ews-build/steps.py]: This actually isn't for debugging, it's because `git` doesn't let me say "add or update" in a single command. If the checkout is already tracking to specified remote, this command will fail. Which is a failure we're interested in ignoring. Open to other ways of expressing that, this was the simplest one I could come up with.

**Review by aj062 — CHANGES_REQUESTED** [review]
This change feels very risky to me for existing patch behaviour. There are major changes like changing ShellCommand to  ShellSequence, using getResultSummary instead of evaluateCommand (note that the purpose of getResultSummary is different, just to provide a text summary of the build, it's being hacked here for doing so much heavy-lifting). If we want to go this route, we need to do a lot of testing, for various corner cases.

Also, there isn't any explicit need to do such large changes to this ApplyPatch step for patch workflow which is working fine. 

It would be nice/simpler if you can:
- add another build-step like CheckOutPRBranch
- add doStepIf (and hideStepIf) to ApplyPatch and CheckOutPRBranch steps so that they are run either on patch or pr appropriately
- update factory to add CheckOutPRBranch just after/before ApplyPatch step

**Comment by JonWBedard** [comment]
> This change feels very risky to me for existing patch behaviour. There are major changes like changing ShellCommand to ShellSequence, using getResultSummary instead of evaluateCommand (note that the purpose of getResultSummary is different, just to provide a text summary of the build, it's being hacked here for doing so much heavy-lifting). If we want to go this route, we need to do a lot of testing, for various corner cases.
> 
> Also, there isn't any explicit need to do such large changes to this ApplyPatch step for patch workflow which is working fine.
> 
> It would be nice/simpler if you can:
> 
> * add another build-step like CheckOutPRBranch
> * add doStepIf (and hideStepIf) to ApplyPatch and CheckOutPRBranch steps so that they are run either on patch or pr appropriately
> * update factory to add CheckOutPRBranch just after/before ApplyPatch step

Using `getResultSummary` was a bit of a hack, new-style buildbot classes don't have `evaluateCommand`, and I think that's because you should override the run function using twisted's defer API (which is was the code I just posted does)

I don't think we should add a new step, I think that will make things more confusing going forward. It seems much better to have a single step that applies whatever code we intend to test. In the current code, `ApplyPatch` should arguably be a `ShellSequence` anyways since one of the ways to apply the path is run `curl` before applying the patch.

File: Tools/CISupport/ews-build/steps.py [hunk 3]
```diff
@@ -565,6 +568,45 @@ def evaluateCommand(self, cmd):
...
+        ]:
+            self.commands.append(util.ShellArg(command=command, logname='stdio', haltOnFailure=True))
+        rc = yield steps.ShellSequence.run(self)
+        if rc == FAILURE:
+            self.build.buildFinished(['Failed to checkout pull request from PR {}'.format(self.getProperty('github.number'))], FAILURE)
+        return rc
+
+    def getResultSummary(self):
+        if self.results != SUCCESS:
+            return {'step': 'failed to checkout branch from pull-request'}
```
Comment (aj062) [inline Tools/CISupport/ews-build/steps.py]: Nit: 'failed' => 'Failed'

File: Tools/CISupport/ews-build/steps.py [hunk 4]
```diff
@@ -565,6 +568,45 @@ def evaluateCommand(self, cmd):
...
+            ['/bin/sh', '-c', 'git remote add {} {}{}.git & true'.format(remote, GITHUB_URL, project)],
+            ['git', 'remote', 'set-url', remote, '{}{}.git'.format(GITHUB_URL, project)],
+            ['git', 'fetch', remote],
+            ['git', 'branch', '-f', branch, 'remotes/{}/{}'.format(remote, branch)],
+            ['git', 'checkout', branch],
+        ]:
+            self.commands.append(util.ShellArg(command=command, logname='stdio', haltOnFailure=True))
+        rc = yield steps.ShellSequence.run(self)
+        if rc == FAILURE:
+            self.build.buildFinished(['Failed to checkout pull request from PR {}'.format(self.getProperty('github.number'))], FAILURE)
```
Comment (aj062) [inline Tools/CISupport/ews-build/steps.py]: maybe "checkout pull request from PR" => "checkout branch from PR"

File: Tools/CISupport/ews-build/steps.py [hunk 5]
```diff
@@ -565,6 +568,45 @@ def evaluateCommand(self, cmd):
         return rc
 
 
+class CheckOutPullRequest(steps.ShellSequence):
+    name = 'checkout-pull-request'
+    description = ['checking-out-pull-request']
+    descriptionDone = ['Checked out pull request']
+    haltOnFailure = True
```
Comment (aj062) [inline Tools/CISupport/ews-build/steps.py]: fine for now, but haltOnFailure would need to be false for Commit-Queue. 
Also you have both haltOnFailure=True as well as custom self.build.buildFinished in run().

**Review by aj062 — COMMENTED** [review]
Looks good on high level. I believe factories unit-tests would need to be updated.

File: Tools/CISupport/ews-build/steps.py [hunk 6]
```diff
@@ -543,9 +546,11 @@ def start(self):
         d.addCallback(lambda res: shell.ShellCommand.start(self))
 
     def hideStepIf(self, results, step):
-        return results == SUCCESS and self.getProperty('sensitive', False)
+        return not self.doStepIf(None) or (results == SUCCESS and self.getProperty('sensitive', False))
```
Comment (aj062) [inline Tools/CISupport/ews-build/steps.py]: Should it be self.doStepIf(None) or self.doStepIf(step)

File: Tools/CISupport/ews-build/steps.py [hunk 7]
```diff
@@ -565,6 +570,51 @@ def evaluateCommand(self, cmd):
...
+            ['/bin/sh', '-c', 'git remote add {} {}{}.git & true'.format(remote, GITHUB_URL, project)],
+            ['git', 'remote', 'set-url', remote, '{}{}.git'.format(GITHUB_URL, project)],
+            ['git', 'fetch', remote],
+            ['git', 'branch', '-f', branch, 'remotes/{}/{}'.format(remote, branch)],
+            ['git', 'checkout', branch],
+        ]:
+            self.commands.append(util.ShellArg(command=command, logname='stdio', haltOnFailure=True))
+
+        rc = yield super(CheckOutPullRequest, self).run()
+        if rc == FAILURE:
```
Comment (aj062) [inline Tools/CISupport/ews-build/steps.py]: Do we need it if since you have haltOnFailure=True?

File: Tools/CISupport/ews-build/steps.py [hunk 8]
```diff
@@ -565,6 +570,51 @@ def evaluateCommand(self, cmd):
...
+            self.build.buildFinished(['Failed to checkout branch from PR {}'.format(self.getProperty('github.number'))], FAILURE)
+        return rc
+
+    def getResultSummary(self):
+        if self.results == SKIPPED:
+            return {'step': 'No pull request to checkout'}
+        if self.results != SUCCESS:
+            return {'step': 'Failed to checkout branch from pull request'}
```
Comment (aj062) [inline Tools/CISupport/ews-build/steps.py]: Might want to print PR number here as this would also go in build summary. Previous statement reads better:
`'Failed to checkout branch from PR {}'.format(self.getProperty('github.number'))`

File: Tools/CISupport/ews-build/steps.py [hunk 9]
```diff
@@ -543,9 +546,11 @@ def start(self):
         d.addCallback(lambda res: shell.ShellCommand.start(self))
 
     def hideStepIf(self, results, step):
-        return results == SUCCESS and self.getProperty('sensitive', False)
+        return not self.doStepIf(None) or (results == SUCCESS and self.getProperty('sensitive', False))
 
     def getResultSummary(self):
+        if self.results == SKIPPED:
+            return {'step': 'No patch to apply'}
```
Comment (aj062) [inline Tools/CISupport/ews-build/steps.py]: Can reword something like: 'Skipping applying patch' or "Skipping applying patch since patch_id isn't provided" or "Skipping applying patch in PR workflow".

'No patch to apply' feels like indicating an error in which patch was expected but not found.

**Review by aj062 — APPROVED** [review]

File: Tools/CISupport/ews-build/steps.py [hunk 10]
```diff
@@ -565,6 +570,51 @@ def evaluateCommand(self, cmd):
...
+
+    @defer.inlineCallbacks
+    def run(self):
+        self.commands = []
+
+        remote = self.getProperty('github.head.repo.full_name', 'origin').split('/')[0]
+        project = self.getProperty('github.head.repo.full_name', self.getProperty('project'))
+        branch = self.getProperty('github.head.ref', 'main')
+
+        for command in [
```
Comment (aj062) [inline Tools/CISupport/ews-build/steps.py]: Can consider combining commands from `clean-and-update-working-directory` step here, and skipping that step in PR workflow, but that can be done later as well. Please atleast include a FIXME for removing that step later-on.

**Comment by aj062** [comment]
Please make sure to deploy and test your latest changes from this PR on uat instance before landing.

File: Tools/CISupport/ews-build/steps.py (hunk 6 above)
Comment (JonWBedard) [reply Tools/CISupport/ews-build/steps.py]: Didn't even notice we had the `step` variable there, definitely `self.doStepIf(step)`

File: Tools/CISupport/ews-build/steps.py (hunk 7 above)
Comment (JonWBedard) [reply Tools/CISupport/ews-build/steps.py]: Good point. I suppose not.

**Comment by JonWBedard** [comment]
Landed https://commits.webkit.org/246230@main (b28a364ba6da2297d556b9aecd2eb32d181b5331)!
